streams/src/main/java/org/apache/kafka/streams/state/KeyValueStore.java

```diff
@@ -32,7 +33,17 @@
  * @param <V> The value type
  */
 @InterfaceStability.Unstable
-public interface KeyValueStore<K, V> extends StateStore {
+public abstract class KeyValueStore<K, V> implements StateStore {
+
+    private boolean checked = false;
+
+    public void checkKeyIsArray(K key) {
+        if (!checked && key.getClass().isArray()) {
+            throw new StreamsException("KeyValueStore cannot support array keys.");
+        } else {
+            checked = true;
+        }
+    }
```

Comment:
@ijuma I think this is a tough call... It's true, Java/Scala collections don't do this... and bugs crop up very frequently as a result. (Observe: this very PR is responding to an example of such a bug!) This turns out to be a very easy mistake to make.

I also wonder if the Array.hashCode/equals surprise isn't even more insidious here than with the standard collection-classes, because these Stores are designed to feel like databases rather than `Collections`, and they're also probably more likely to contain byte[] keys than your typical HashMap.

So, yes, this could be addressed by documentation, but I think the one-time validation might be worthwhile to prevent any accidents from making it into production.


All that being said, I also wonder if moving the serde to be applied _in front of_ the Maps/Sets inside these classes might be a better approach. If the stores simply always operated on ByteBuffer (or whatever) internally, then there'd be no surprises.